File: src/main/java/org/noxet/noxetserver/playerdata/types/PDTMapStringMapStringLocation.java
package org.noxet.noxetserver.playerdata.types;

import org.bukkit.Location;
import org.bukkit.configuration.MemorySection;
import org.bukkit.configuration.file.YamlConfiguration;
import org.noxet.noxetserver.playerdata.PlayerDataType;

import java.util.HashMap;
import java.util.Map;

public class PDTMapStringMapStringLocation implements PlayerDataType<Map<String, Map<String, Location>>> {
    @Override
    public Map<String, Map<String, Location>> getEmptyValue() {
        return new HashMap<>();
    }

    @Override
    @SuppressWarnings("unchecked")
    public Map<String, Map<String, Location>> getValue(YamlConfiguration config, String key) {
        Map<String, Map<String, Location>> result = new HashMap<>();

        MemorySection memorySection = (MemorySection) config.get(key);
        assert memorySection != null;

        for(Map.Entry<String, Object> topEntries : memorySection.getValues(false).entrySet())
            result.put(topEntries.getKey(), (Map<String, Location>) topEntries.getValue());

        return result;
    }
}

File: src/main/java/org/noxet/noxetserver/playerdata/types/PDTMapStringMapStringLocation.scala
package org.noxet.noxetserver.playerdata.types

import org.bukkit.Location
import org.bukkit.configuration.MemorySection
import org.bukkit.configuration.file.YamlConfiguration
import org.noxet.noxetserver.playerdata.PlayerDataType

import scala.jdk.CollectionConverters.*

class PDTMapStringMapStringLocation extends PlayerDataType[Map[String, Map[String, Location]]]:
  override def getEmptyValue: Map[String, Map[String, Location]] = Map.empty

  override def getValue(config: YamlConfiguration, key: String): Map[String, Map[String, Location]] =
    Option(config.get(key).asInstanceOf[MemorySection]) match
      case Some(memSec) =>
        memSec.getValues(false).asScala.toMap
          .map((name, memSec) => name ->
            memSec.asInstanceOf[MemorySection]
              .getValues(false).asScala
              .asInstanceOf[Map[String, Location]])
      case None => Map.empty